FEATURE: Add new group visibility option for "logged on users"

pmusaraj opened the pull request:
In some cases it is useful to only allow logged on users to be able to view group membership. That's what this PR does. It also sets all automatic groups (except`everyone`) to be visible to "logged on users", previously they were marked as public but suppressed in the group page for non-staff.  

Admin UI screenshot: 
<img width="489" alt="image" src="https://user-images.githubusercontent.com/368961/60209925-ad4b8700-9829-11e9-8c5f-c010a0eb971f.png">

SamSaffron commented:
This looks good to me, go ahead and merge... a tiny bit mixed about id (4) belonging to this thing, cause the numbers are kind of expanding in scope... but re-numbering this thing is going to be complicated. 

Can you first perhaps see how complex it will be to re-number?

pmusaraj commented:
It shouldn't be too painful to renumber. Is this below the preferred numbering? 

public: 0,
logged_on_users: 1,
members: 2,
staff: 3,
owners: 4

SamSaffron commented:
Yeah I think that numbering is right ... technically "staff" kind of means "moderators" cause "admins" unconditionally get to see groups. I think "owners" can be 4 as long as "mods" can not see it when set... otherwise it needs to be 3.

pmusaraj commented:
I have added a migration for the renumbering, and made two more changes.

First, I set the default visibility for the "everyone" group to be "staff" in `db/fixtures/002_groups.rb`. This way it's more consistent with the `Group.refresh_automatic_group!` method that gets called by the ensure_consistency job, see https://github.com/discourse/discourse/blob/master/app/models/group.rb#L320 

Second change: I modified the language in the UI, adding a field description with "Admins can see all groups.": 

<img width="305" alt="image" src="https://user-images.githubusercontent.com/368961/60740980-cfce5600-9f35-11e9-88ed-37375bf83d60.png">

and removed the "admins" from the label of each option in the dropdown: 

<img width="691" alt="image" src="https://user-images.githubusercontent.com/368961/60740989-d8bf2780-9f35-11e9-8258-f171e846de37.png">

eviltrout commented on db/migrate/20190705173948_renumber_group_visibility_levels.rb:
```diff
@@ -0,0 +1,16 @@
+
+class RenumberGroupVisibilityLevels < ActiveRecord::Migration[5.2]
+  def up
+    execute "UPDATE groups SET visibility_level = 4 WHERE visibility_level = 3"
```
Super minor but this could have been `SET visibility_level = visibility_level + 1` right? And below the opposite? (`-1`).
No need to change it just an observation :)

pmusaraj replied:
Yeah, that would have been better (I would need to modify it so that 0-level groups shouldn't be bumped to level 1 on `up`).

eviltrout commented on spec/components/guardian_spec.rb:
```diff
@@ -3144,9 +3167,10 @@
+      expect(Guardian.new(another_user).can_see_groups?([group])).to eq(false)
     end
 
-    it 'correctly handles member visibile groups' do
```
Oh my this typo is in so many places. Just goes to show how much code is copied and pasted :)
